crates/turborepo-telemetry/src/lib.rs

```diff
@@ -0,0 +1,338 @@
+    /// within that timeout, it will log an error and drop the handle.
+    pub async fn close_with_timeout(self) {
+        if let Err(err) = tokio::time::timeout(EVENT_TIMEOUT, self.close()).await {
+            debug!("failed to close telemetry handle. error: {}", err)
+        } else {
+            debug!("telemetry handle closed")
+        }
+    }
+}
+
+struct Worker<C> {
+    rx: mpsc::UnboundedReceiver<TelemetryEvent>,
+    buffer: Vec<TelemetryEvent>,
+    senders: FuturesUnordered<JoinHandle<()>>,
+    // Used to cancel the worker
+    exit_ch: oneshot::Sender<()>,
+    client: C,
+    telemetry_id: String,
+    session_id: String,
+    enabled: bool,
+    ui: UI,
+}
+
+impl<C: telemetry::TelemetryClient + Clone + Send + Sync + 'static> Worker<C> {
+    pub fn start(mut self) -> JoinHandle<()> {
+        tokio::spawn(async move {
+            let mut timeout = tokio::time::sleep(NO_TIMEOUT);
+            loop {
+                select! {
+                    // We want the events to be prioritized over closing
+                    biased;
+                    event = self.rx.recv() => {
+                        if let Some(event) = event {
+                            self.buffer.push(event);
+                        } else {
+                            // There are no senders left so we can shut down
+                            break;
+                        }
+                        if self.buffer.len() == BUFFER_THRESHOLD {
+                            self.flush_events();
+                            timeout = tokio::time::sleep(NO_TIMEOUT);
+                        } else {
+                            timeout = tokio::time::sleep(EVENT_TIMEOUT);
+                        }
+                    }
+                    _ = timeout => {
+                        self.flush_events();
+                        timeout = tokio::time::sleep(NO_TIMEOUT);
+                    }
+                    _ = self.exit_ch.closed() => {
+                        break;
+                    }
+                }
+            }
+            self.flush_events();
+            while let Some(result) = self.senders.next().await {
+                if let Err(err) = result {
+                    debug!("failed to send telemetry event. error: {}", err)
+                }
+            }
+        })
+    }
+
+    pub fn flush_events(&mut self) {
+        if !self.buffer.is_empty() {
+            let events = std::mem::take(&mut self.buffer);
+            debug!(
+                "Starting telemetry event queue flush (num_events={:?})",
+                events.len()
+            );
+            let handle = self.send_events(events);
+            self.senders.push(handle);
+            debug!("Done telemetry event queue flush");
+        }
+    }
+
+    fn send_events(&self, events: Vec<TelemetryEvent>) -> JoinHandle<()> {
+        if !self.enabled {
+            // do nothing
+            return tokio::spawn(async move {});
+        }
```

Comment:
Tokio tasks are cheap, but `None` is cheaper. 

Will need to update callers to only add present join handles to the senders list.
```suggestion
    fn send_events(&self, events: Vec<TelemetryEvent>) -> Option<JoinHandle<()>> {
        if !self.enabled {
            // do nothing
            return None;
        }
```

Reply:
Love it.